### PR TITLE
fix: migrate config fields to correct the location 

### PR DESCRIPTION
**Pull request by ImCesar**
- Closes [UNIFY-957](https://cypress-io.atlassian.net/browse/UNIFY-957)

Edit(lachlan): I added a bunch of example projects w/ descriptions, [example](https://github.com/cypress-io/cypress/pull/19940/files#diff-de569ec304196a4d4070ed3a4eb3e868ad45245586dd149c272dd417b87dee5c). I added [`assertSuccessfulMigration`](https://github.com/cypress-io/cypress/pull/19940/files#diff-2e929d53cafe06c7294f6de9bb34226e359f2a4acaac297b87d4563646b46959R45) to assert the generated `cypress.config.js` is correct at the end of each migration. It asserts against the `expected-cypress.config.js` in the test project's root. 

Edit2(lachlan): many example projects are added, in `system-tests/projects`. They are named `migration-<name>`. Each has a `README.md` explaining what scenario is represents, and a `expected-cypress.config.js`, which we use to verify the migration is successful by comparing it to the generated `cypress.config.js`.

Screen Shot 2022-02-01 at 3 53 37 pm

### Additional details
Follows different edge cases for how to transfer different config fields

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/data-context/src/util/migration.ts [hunk 1]
```diff
@@ -303,3 +271,69 @@ export function renameSpecPath (spec: string) {
...
+      }
+    }
+
+    return { ...acc, global: { ...acc.global, [key]: val } }
+  }, { global: {}, e2e: {}, component: {} })
+}
+
+function getSpecPattern (cfg: OldCypressConfig, testType: TestingType) {
+  const specPattern = cfg[testType]?.testFiles ?? cfg.testFiles ?? '**/*.cy.js'
+  const customComponentFolder = cfg.component?.componentFolder ?? cfg.componentFolder ?? null
```
Comment (lmiller1990) [inline packages/data-context/src/util/migration.ts:326]: Please use this function https://github.com/cypress-io/cypress/blob/10.0-release/packages/data-context/src/sources/migration/shouldShowSteps.ts#L48 it has the `false` edge case

File: packages/data-context/src/util/migration.ts [hunk 2]
```diff
@@ -303,3 +271,69 @@ export function renameSpecPath (spec: string) {
...
+        e2e: { ...acc.e2e, [key]: val },
+      }
+    }
+
+    return { ...acc, global: { ...acc.global, [key]: val } }
+  }, { global: {}, e2e: {}, component: {} })
+}
+
+function getSpecPattern (cfg: OldCypressConfig, testType: TestingType) {
+  const specPattern = cfg[testType]?.testFiles ?? cfg.testFiles ?? '**/*.cy.js'
```
Comment (lmiller1990) [inline packages/data-context/src/util/migration.ts:325]: We have a function for this: https://github.com/cypress-io/cypress/blob/cec46dfc058a8745438060552729f158abc3f9a1/packages/data-context/src/sources/migration/shouldShowSteps.ts#L24-L28

File: packages/data-context/src/util/migration.ts [hunk 3]
```diff
@@ -303,3 +271,69 @@ export function renameSpecPath (spec: string) {
...
+  const customComponentFolder = cfg.component?.componentFolder ?? cfg.componentFolder ?? null
+
+  if (testType === 'component' && customComponentFolder) {
+    return `${customComponentFolder}/${specPattern}`
+  }
+
+  const customIntegrationFolder = cfg.e2e?.integrationFolder ?? cfg.integrationFolder ?? null
```
Comment (lmiller1990) [inline packages/data-context/src/util/migration.ts:332]: https://github.com/cypress-io/cypress/blob/cec46dfc058a8745438060552729f158abc3f9a1/packages/data-context/src/sources/migration/shouldShowSteps.ts#L40

**Review by lmiller1990 — COMMENTED** [review]
Left some comments!

File: packages/data-context/src/util/migration.ts (hunk 2 above)
Comment (ImCesar) [reply packages/data-context/src/util/migration.ts:326]: I may be mistake but I don't want the actual files, I want the testFiles pattern

File: packages/data-context/src/util/migration.ts (hunk 3 above)
Comment (ImCesar) [reply packages/data-context/src/util/migration.ts:338]: I added these back because they kept giving me the default. Though I'm not 100% on what we should do if one isn't provided. Should it be `${testFiles}` or should it be `${defaultFolder}/${testFiles}`?

**Review by lmiller1990 — DISMISSED** [review]
Pushed some code - we have pretty good coverage now, but def need a second set of eyes.

**Review by elevatebart — DISMISSED** [review]

**Review by lmiller1990 — APPROVED** [review]

File: packages/launchpad/src/migration/ConvertConfigFile.vue [hunk 4]
```diff
@@ -146,30 +149,40 @@ import MigrationListItem from './fragments/MigrationListItem.vue'
...
+fragment ConvertConfigFile on Query {
+  migration {
+    configBeforeCode
+    configAfterCode
+    hasCustomIntegrationFolder
+    hasCustomIntegrationTestFiles
+    hasCustomComponentFolder
+    hasCustomComponentTestFiles
+  }
+  currentProject {
```
Comment (elevatebart) [inline packages/launchpad/src/migration/ConvertConfigFile.vue:161]: We should add a hastypescript flag to the migration member to avoid this pain.

**Review by elevatebart — APPROVED** [review]

**Review by lmiller1990 — DISMISSED** [review]

**Review by elevatebart — APPROVED** [review]

File: system-tests/projects/migration-component-testing-customized/README.md [hunk 5]
```diff
@@ -1,4 +1,8 @@
-A component tesing project without e2e. We ask the users to migrate their specs manually. `supportFile` is `false`.
+# Migration Component Testing Customized
```
Comment (tbiethman) [inline system-tests/projects/migration-component-testing-customized/README.md:1]: These README's are super helpful :+1:

**Review by tbiethman — APPROVED** [review]

**Review by lmiller1990 — DISMISSED** [review]

**Review by elevatebart — DISMISSED** [review]

**Review by tbiethman — APPROVED** [review]

**Review by lmiller1990 — APPROVED** [review]

**Review by elevatebart — APPROVED** [review]

**Review by tbiethman — APPROVED** [review]
